cirq-core/cirq/ops/control_values.py

```diff
@@ -38,150 +42,179 @@ class AbstractControlValues(abc.ABC):
-        [0, qid_shapes) for each qubit.
-        """
+        Args:
+          other: An instance of `AbstractControlValues`.
 
-    @abc.abstractmethod
-    def _are_ones(self) -> bool:
-        """Checks whether all control values are equal to 1."""
+        Returns:
+          An instance of `AbstractControlValues` that represents the cartesian product of
+          control values represented by `self` and `other`.
+        """
+        return SumOfProducts(
+            tuple(x + y for (x, y) in itertools.product(self.expand(), other.expand()))
+        )
 
-    @abc.abstractmethod
-    def _json_dict_(self) -> Dict[str, Any]:
-        pass
+    def __or__(self, other: 'AbstractControlValues') -> 'AbstractControlValues':
+        """Returns a union of all control values predicates in `self` + `other`.
 
-    def __iter__(self) -> Generator[Tuple[int, ...], None, None]:
-        for assignment in self._expand():
-            yield assignment
+        Both `self` and `other` must represent control values for the same set of qubits and
+        hence their `or` would also be a control value object acting on the same set of qubits.
+        For example:
 
-    def __eq__(self, other) -> bool:
-        """Returns True iff self and other represent the same configurations.
+        >>> cv1 = cirq.ProductOfSums([(0, 1), 2])
+        >>> cv2 = cirq.SumOfProducts([[0, 0], [1, 1]])
+        >>> assert cirq.num_qubits(cv1 | cv2) == cirq.num_qubits(cv1) == cirq.num_qubits(cv2)
 
         Args:
-            other: A AbstractControlValues object.
+          other: An instance of `AbstractControlValues`.
 
         Returns:
-            boolean whether the two objects are equivalent or not.
+          An instance of `AbstractControlValues` that represents the union of control values
+          represented by `self` and `other`.
+
+        Raises:
+            ValueError: If `cirq.num_qubits(self) != cirq.num_qubits(other)`.
         """
-        if not isinstance(other, AbstractControlValues):
-            other = ProductOfSums(other)
-        return sorted(v for v in self) == sorted(v for v in other)
+        if protocols.num_qubits(self) != protocols.num_qubits(other):
+            raise ValueError(
+                f"Control values {self} and {other} must act on equal number of qubits"
+            )
+        return SumOfProducts(tuple(x for x in [*self.expand(), *other.expand()]))
 
 
-@dataclass(frozen=True, eq=False)
 class ProductOfSums(AbstractControlValues):
-    """ProductOfSums represents control values in a form of a cartesian product of tuples."""
+    """Represents control values as N OR (sum) clauses, each of which applies to one qubit."""
 
-    _internal_representation: Tuple[Tuple[int, ...], ...]
+    def __init__(self, data: Sequence[Union[int, Collection[int]]]):
+        self._qubit_sums: Tuple[Tuple[int, ...], ...] = tuple(
+            (cv,) if isinstance(cv, int) else tuple(sorted(frozenset(cv))) for cv in data
```

Comment:
nit: don't need `frozenset` here since it's just a temporary value anyway:
```suggestion
            (cv,) if isinstance(cv, int) else tuple(sorted(set(cv))) for cv in data
```

Reply:
done